### DISCUSSION TITLE
Using Typescript SDK to pass Optional object of type fungible metadata

### DISCUSSION THREAD

**Discussion by KodamaHQ** [aptos-labs/aptos-developer-discussions#549] [Questions] [answered]
### Discord user ID

@XYZ

### Describe your question in detail.

I want to call a public script function using typescript sdk. 

public entry fun test(&signer, Optional<0x1::object::Object<0x1::fungible_asset::Metadata>>,  u64)

what typescript object would be a
0x1::object::Object<0x1::fungible_asset::Metadata>

or in case i dont want it an empty optional. 

I saw these are the script function argument types: 
    //   type ScriptFunctionArgumentTypes = Bool | U8 | U16 | U32 | U64 | U128 | U256 | AccountAddress | MoveVector<ScriptFunctionArgumentTypes> | MoveString | FixedBytes | Serialized;

Saw MoveOption is missing the serializeForScriptFunction method. 
any idea how to work around it ? 

### What have you tried or looked at?  Or how can we reproduce the error?

```typescript 
 let ty_args: Array<TypeTag> = [];
 let args: Array<TransactionArgument> = [new U64(1000)];
ty_args.push(new TypeTagU8())
    args.push(new U64(BigInt(27)))
    args.push(new Bool(false))
    args.push(AccountAddress.fromStringStrict("0x1"))

/* Here I want to get an optional object which is a meta data of a fungible asset. and no idea how. */
    args.push(new MoveString(""))

    const payload = new TransactionPayloadScript(
        new Script(
        Hex.fromHexString(script.parsed).toUint8Array(),
    ty_args ,args ));
```

### Which operating system are you using?

Linux (Ubuntu, Fedora, Windows WSL, etc.)

### Which SDK or tool are you using? (if any)

TypeScript SDK

**Comment by Pcharlesme** [comment]
To pass an Optional object in the TypeScript SDK when calling a Move script function, you need to use MoveOption. While it's true that MoveOption is missing the serializeForScriptFunction method, you can work around this limitation by serializing the optional value directly using a custom approach.

**Comment by Pcharlesme** [comment, marked as answer]
You can check out the Example below to pass an Optional argument of type 0x1::object::Object<0x1::fungible_asset::Metadata> or an empty Optional:

`import { MoveOption, TransactionArgument, TransactionPayloadScript, TypeTagStruct, TypeTagAddress, TypeTagVector, U64, AccountAddress } from "@aptos-labs/aptos-ts-sdk";
import { HexString } from "@aptos-labs/hex-string";

// Helper to create a MoveOption for Optional<Metadata>
function createOptionalMetadata(metadata: any | null): MoveOption {
    if (metadata === null) {
        return new MoveOption(null); // Represents an empty Optional
    } else {
        // Serialize the metadata object manually
        const serializedMetadata = serializeMetadata(metadata);
        return new MoveOption(serializedMetadata);
    }
}

 
function serializeMetadata(metadata: any): Uint8Array {
    // Assuming metadata has fields { name: string, description: string, decimals: number }
    const nameBytes = new TextEncoder().encode(metadata.name);
    const descriptionBytes = new TextEncoder().encode(metadata.description);
    const decimalsBytes = new Uint8Array([metadata.decimals]);

    // Combine all fields (you may need more robust serialization based on metadata structure)
    return new Uint8Array([...nameBytes, ...descriptionBytes, ...decimalsBytes]);
}

// Example script function call
const tyArgs: Array<TypeTag> = [
    new TypeTagStruct(new StructTag(AccountAddress.fromHex("0x1"), "fungible_asset", "Metadata", [])),
];
const args: Array<TransactionArgument> = [];

// Signer
args.push(new AccountAddress("0x1"));

 
const optionalMetadata = createOptionalMetadata(null);
args.push(optionalMetadata);

 
args.push(new U64(1000));

// Create transaction payload
const payload = new TransactionPayloadScript(
    new Script(
        HexString.fromHex("your_script_hex").toUint8Array(),
        tyArgs,
        args
    )
);

 
`

**Comment by Pcharlesme** [comment]
@ranlavanet

**Reply by KodamaHQ** [reply]
Wow thats amazing. thank you so much!
